### PR TITLE
Add `cupy.poly`

### PR DESCRIPTION
**Pull request by Dahlia-Chehata**
blocked by #3828

### CODE REVIEW

File: cupy/lib/polynomial.pyx [hunk 1]
```diff
@@ -4,6 +4,47 @@ import cupy
...
+    else:
+        raise ValueError('Input must be 1d or non-empty square 2d array.')
+
+    if seq.size == 0:
+        return 1.0
+
+    out = cupy.ones((1,), seq.dtype)
+    seq = cupy.column_stack((cupy.ones(seq.size, seq.dtype), -seq))
+    for zero in seq:
+        out = cupy.convolve(out, zero, 'full')
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:44]: Please parallelize this loop. We can achieve this computation with O(log n) time FFT call.

Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:44]: The best algorithm I know for coefficients computation using FFT is a recursive O(n log^2 n) time one. Is that what you mean or do you have a better approach?

Comment (asi1024) [reply cupy/lib/polynomial.pyx:44]: Yes.

File: cupy/lib/polynomial.pyx [hunk 2]
```diff
@@ -4,6 +4,16 @@ import cupy
...
 
 
+def _fft_poly(seq):
+    n = seq.shape[0]
+    if n == 1:
+        return seq[0]
+    if n == 2:
+        return cupy.math.misc._fft_convolve(seq[0], seq[1], 'full')
+    return cupy.math.misc._fft_convolve(_fft_poly(seq[:n/2]),
+                                        _fft_poly(seq[n/2:]), 'full')
```
Comment (asi1024) [inline cupy/lib/polynomial.pyx:14]: This implementation calls FFT O(n) times, but we can achieve O(log n) times if we support Nd convolve.

Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx:14]: You suggest supporting scipy.signal.convolve or extending the functionality of cupy.convolve ?

Comment (asi1024) [reply cupy/lib/polynomial.pyx:14]: OK!

**Comment by kmaehashi** [comment]
Related discussion: https://github.com/cupy/cupy/pull/3371#issuecomment-654464151

File: cupy/lib/polynomial.pyx (hunk 1 above)
Comment (Dahlia-Chehata) [reply cupy/lib/polynomial.pyx]:
```
def _fft_poly(seq):
    n = seq.shape[0]
    if n == 1:
        return seq[0]
    if n == 2:
        return cupy.math.misc._fft_convolve(seq[0], seq[1], 'full')
    return signal.convolve(_fft_poly(seq[:n / 2]), _fft_poly(seq[n / 2:]), 'full', 'direct')
```
I'm not really sure if that's what you meant or not. If so, I want to clarify that
1. complex dtype give different results in comparison with numpy's
2. Only `direct` method is currently supported for `cupyx.scipy.signal.convolve` that's why I have to specify the method.

Comment (asi1024) [reply cupy/lib/polynomial.pyx]: No, it requires O(n) time FFT kernel call.

**Comment by asi1024** [comment]
@Dahlia-Chehata Could you add TODO comment for the issue https://github.com/cupy/cupy/pull/3547#discussion_r449964352, or support 2d inputs in `cupyx.scipy.signal.convolve` and use it?

**Comment by asi1024** [comment]
@Dahlia-Chehata Can I takeover the PR?

**Pull request closed without merging**